pages/blog/api-first-with-asyncapi.md

```diff
@@ -0,0 +1,284 @@
+canonical: https://zenwave360.github.io/Event-Driven-Architectures/API-First-with-AsyncAPI
+authors:
+  - name: Ivan Garcia Sainz-Aja
+    photo: /img/avatars/ivangsa.webp
+    link: https://www.linkedin.com/in/ivangarciasainzaja/
+    byline: Software Architect at Sngular. Speaker. Building ZenWave 360º
+excerpt: 'If you are familiar with OpenAPI and OpenAPI Generator API-First process, doing API-First with AsyncAPI is a similar process.'
+---
+
+# API-First with AsyncAPI
+
+- [API-First with AsyncAPI](#api-first-with-asyncapi)
+  - [Broker-based APIs are Symmetric](#broker-based-apis-are-symmetric)
+  - [Events, Commands, and Messages](#events-commands-and-messages)
+  - [Understanding AsyncAPI Definition](#understanding-asyncapi-definition)
+    - [Info](#info)
+    - [Servers](#servers)
+    - [Channels: Publish / Subscribe](#channels-publish--subscribe)
+    - [Messages](#messages)
+    - [Message Payloads / Schemas](#message-payloads--schemas)
+    - [Reusing Configurations: Operation Traits, Message Traits...](#reusing-configurations-operation-traits-message-traits)
+  - [Different Styles of Event Messages](#different-styles-of-event-messages)
+    - [Notification Messages](#notification-messages)
+    - [State Transfer Messages](#state-transfer-messages)
+    - [Domain Event Messages](#domain-event-messages)
+  - [Next: Java Code Generator for AsyncAPI](#next-java-code-generator-for-asyncapi)
+
+If you are familiar with OpenAPI and OpenAPI Generator API-First workflow:
+
+- First write the OpenAPI definition, collaborating between API providers and API consumers.
+- Then use OpenAPI Generator, either the maven plugin or a CLI, to generate some DTOs and interfaces from your OpenAPI definition.
+- Implementing the generated interfaces you can create a service for the API.
+- As a client, you can use generated the interfaces to consume the API with some HTTP client generated behind the scenes.
+
+When doing API-First with AsyncAPI, the process is similar. After you generate some interfaces and DTOs from your API definition, you use the generated interfaces to produce messages, sending them to the broker, and implement them to consume messages from the broker.
+
+There is still a fundamental difference between OpenAPI and AsyncAPI: OpenAPI is used to document Request-Response / Client-Server APIs, while AsyncAPI is used to document Event-Driven APIs which, except for websockets, are Broker-based. 
+
+And broker-based APIs, unlike Client-Server, are inherently **symmetric**.
+
+## Broker-based APIs are Symmetric
+
+![Client-server vs broker-based EDAs](/img/posts/zenwave/client-server-vs-broker-eda.excalidraw.svg)
+
+Because APIs mediated by a broker are inherently **symmetric** it's difficult to establish the roles of client/server: what represents a `publish` operation from one side will be a `subscribe` operation seen from the other side. Also, a given service can act as a publisher and subscriber on the same API.
```

Comment:
```suggestion
Because APIs mediated by a broker are inherent **symmetric**, it's difficult to establish the roles of the client/server: what represents a `publish` operation from one side will be a `subscribe` operation seen from the other side. Also, a given service can act as a publisher and subscriber on the same API.
```